File: linux/native/src/helper/install_strategy.cc
#include "install_strategy.h"

#include <algorithm>
#include <set>
#include <utility>

namespace desktop_updater::helper {
namespace {

const std::set<std::pair<std::string, std::string>> kLinuxCapabilities = {
    {"directoryReplace", "platformDirectory"},
    {"singleFileReplace", "platformFile"},
    {"systemPackageTransaction", "apt"},
    {"systemPackageTransaction", "dnf"},
    {"externalManagedRefresh", "flatpak"},
    {"externalManagedRefresh", "snap"},
};

}  // namespace

LinuxStrategyCapability SelectLinuxInstallStrategy(
    const std::vector<LinuxStrategyCapability>& policy_capabilities,
    const std::vector<LinuxStrategyCapability>& protocol_capabilities,
    const LinuxStrategyRequest& request) {
  const std::pair<std::string, std::string> requested{request.strategy,
                                                      request.provider};
  if (kLinuxCapabilities.count(requested) == 0) {
    throw LinuxInstallStrategyError("unsupported strategy/provider pair");
  }
  if (!request.caller_arguments.empty()) {
    throw LinuxInstallStrategyError("caller_arguments are never executable");
  }
  if (request.direct_revision_mutation) {
    throw LinuxInstallStrategyError("direct managed revision mutation rejected");
  }
  if (request.dangerous_sideload) {
    throw LinuxInstallStrategyError("production dangerous sideload rejected");
  }
  if (request.strategy == "singleFileReplace" && request.root_owned_target &&
      !request.broker_authenticated) {
    throw LinuxInstallStrategyError(
        "root-owned AppImage requires authenticated installed broker");
  }
  const auto policy = std::find_if(
      policy_capabilities.begin(), policy_capabilities.end(),
      [&](const auto& capability) {
        return capability.strategy == request.strategy &&
               capability.provider == request.provider;
      });
  if (policy == policy_capabilities.end()) {
    throw LinuxInstallStrategyError("sealed policy denied capability");
  }
  const LinuxStrategyCapability capability{request.strategy, request.provider};
  if (std::find(protocol_capabilities.begin(), protocol_capabilities.end(),
                capability) == protocol_capabilities.end()) {
    throw LinuxInstallStrategyError("protocol capability missing");
  }
  return capability;
}

LinuxFileTransactionResult ExecuteLinuxDirectoryReplace(
    LinuxFileTransaction& transaction) {
  return transaction.Execute();
}

}  // namespace desktop_updater::helper

File: macos/install_helper/Sources/DesktopUpdaterInstallHelper/InstallStrategy.swift
import Foundation

enum NativeInstallStrategy: String, CaseIterable {
    case directoryReplace
    case singleFileReplace
    case verifiedInstallerHandoff
    case systemPackageTransaction
    case externalManagedRefresh
}

enum NativeInstallProvider: String, CaseIterable {
    case platformDirectory
    case platformFile
    case macosInstaller
    case windowsInno
    case apt
    case dnf
    case flatpak
    case snap
}

struct InstallStrategyCapability: Hashable {
    let strategy: NativeInstallStrategy
    let provider: NativeInstallProvider
}

struct InstallStrategyRequest {
    let strategy: NativeInstallStrategy
    let provider: NativeInstallProvider
    let brokerAuthenticated: Bool
    let targetRootOwned: Bool
    let callerArguments: [String]
    let directRevisionMutation: Bool
    let dangerousSideload: Bool

    init(
        strategy: NativeInstallStrategy,
        provider: NativeInstallProvider,
        brokerAuthenticated: Bool = false,
        targetRootOwned: Bool = false,
        callerArguments: [String] = [],
        directRevisionMutation: Bool = false,
        dangerousSideload: Bool = false
    ) {
        self.strategy = strategy
        self.provider = provider
        self.brokerAuthenticated = brokerAuthenticated
        self.targetRootOwned = targetRootOwned
        self.callerArguments = callerArguments
        self.directRevisionMutation = directRevisionMutation
        self.dangerousSideload = dangerousSideload
    }
}

enum InstallStrategySelectionError: Error, Equatable {
    case unsupportedPair
    case policyDenied
    case protocolCapabilityMissing
    case callerArgumentsRejected
    case directRevisionMutationRejected
    case dangerousSideloadRejected
    case rootFileRequiresBroker
}

struct MacInstallStrategySelector {
    private static let supported: Set<InstallStrategyCapability> = [
        .init(strategy: .directoryReplace, provider: .platformDirectory),
        .init(strategy: .singleFileReplace, provider: .platformFile),
        .init(
            strategy: .verifiedInstallerHandoff,
            provider: .macosInstaller
        ),
    ]

    private let policyCapabilities: Set<InstallStrategyCapability>
    private let protocolCapabilities: Set<InstallStrategyCapability>

    init(
        policyCapabilities: Set<InstallStrategyCapability>,
        protocolCapabilities: Set<InstallStrategyCapability>
    ) {
        self.policyCapabilities = policyCapabilities
        self.protocolCapabilities = protocolCapabilities
    }

    func select(
        _ request: InstallStrategyRequest
    ) throws -> InstallStrategyCapability {
        let requested = InstallStrategyCapability(
            strategy: request.strategy,
            provider: request.provider
        )
        guard Self.supported.contains(requested) else {
            throw InstallStrategySelectionError.unsupportedPair
        }
        guard request.callerArguments.isEmpty else {
            throw InstallStrategySelectionError.callerArgumentsRejected
        }
        guard !request.directRevisionMutation else {
            throw InstallStrategySelectionError
                .directRevisionMutationRejected
        }
        guard !request.dangerousSideload else {
            throw InstallStrategySelectionError.dangerousSideloadRejected
        }
        if request.strategy == .singleFileReplace,
           request.targetRootOwned,
           !request.brokerAuthenticated {
            throw InstallStrategySelectionError.rootFileRequiresBroker
        }
        guard policyCapabilities.contains(requested) else {
            throw InstallStrategySelectionError.policyDenied
        }
        guard protocolCapabilities.contains(requested) else {
            throw InstallStrategySelectionError.protocolCapabilityMissing
        }
        return requested
    }
}

struct MacFileStrategyExecutor {
    func execute(
        capability: InstallStrategyCapability,
        transaction: MacFileTransaction
    ) throws -> MacFileTransactionResult {
        guard capability == .init(
            strategy: .directoryReplace,
            provider: .platformDirectory
        ) || capability == .init(
            strategy: .singleFileReplace,
            provider: .platformFile
        ) else {
            throw InstallStrategySelectionError.unsupportedPair
        }
        return try transaction.execute()
    }
}
